save-preprocessor/src/main/kotlin/com/saveourtool/save/preprocessor/service/TestDiscoveringService.kt

```diff
@@ -186,10 +188,45 @@ class TestDiscoveringService(
      * @throws PluginException if configs use unknown plugin
      */
     @Suppress("UnsafeCallOnNullableType")
-    fun discoverAndSaveAllTests(rootTestConfig: TestConfig, testSuites: List<TestSuite>) = getAllTests(rootTestConfig, testSuites)
+    fun discoverAllTestsIntoMap(
+        rootTestConfig: TestConfig,
+        testSuites: List<TestSuiteDto>,
+    ) = getAllTests(rootTestConfig, testSuites).convertToMap().updatePluginNames()
+
+    @Suppress("TYPE_ALIAS")
+    private fun Mono<Map<TestSuiteDto, List<TestDto>>>.saveTestSuitesAndTests() = flatMap { testsMap ->
+        testsMap.run {
+            saveTestSuites().also {
+                saveTests()
+            }
+        }
+    }
+
+    @Suppress("TYPE_ALIAS")
+    private fun Map<TestSuiteDto, List<TestDto>>.saveTestSuites() = keys.toList().save()
+
+    @Suppress("TYPE_ALIAS")
+    private fun Map<TestSuiteDto, List<TestDto>>.saveTests() = values.flatten()
         .toFlux()
         .save()
 
+    @Suppress("TYPE_ALIAS")
+    private fun Sequence<Pair<TestSuiteDto, TestDto>>.convertToMap() = groupBy({ (testSuite, _) ->
+        testSuite
+    }) { (_, test) ->
+        test
+    }
+
+    @Suppress("TYPE_ALIAS")
+    private fun Map<TestSuiteDto, List<TestDto>>.updatePluginNames() = map { (testSuite, tests) ->
+        val collectedPluginNames = tests.collectPluginNames()
+        log.debug {
+            "Test suite ${testSuite.name} has [$collectedPluginNames] plugins."
+        }
+        testSuite.copy(plugins = collectedPluginNames) to tests
```

Comment:
Why do we need to collect plugins from tests while we have `TestConfig.plugins` in `TestDiscoveringServiceTest.getAllTestSuites`?

Reply:
The idea is to collect reliable information about tests in order to be sure that we will not face empty test suites